pkg/ingester/user_metrics_metadata.go

```diff
@@ -47,15 +47,15 @@ func (mm *userMetricsMetadata) add(metric string, metadata *mimirpb.MetricMetada
 		// Verify that the user can create more metric metadata given we don't have a set for that metric name.
 		if err := mm.limiter.AssertMaxMetricsWithMetadataPerUser(mm.userID, len(mm.metricToMetadata)); err != nil {
 			mm.metrics.discardedMetadataPerUserMetadataLimit.WithLabelValues(mm.userID).Inc()
-			return makeLimitError(mm.limiter.FormatError(mm.userID, err))
+			return formatMaxMetadataPerUserError(mm.limiter, mm.userID)
```

Comment:
A concern I have with this is the implicit assumption that ` mm.limiter.AssertMaxMetricsWithMetadataPerUser` returns `errMaxMetadataPerUserLimitExceeded`. By contract (doc string), it can return "an error", there's no guarantee of the kind of error. I.e., the assumption may break in the future.

Reply:
You're right, there's an implicit assumption there. Good news is that it's not worse than it was before: previous `makeLimitError` also made that implicit assumption (it just assumed that it was one of 4 limit errors, instead of assuming it's a specific one).

I think it's a good moment to solve this. If we take a look into `AssertMaxMetricsWithMetadataPerUser` it actually only returns that error or nil, and this is the only usage of that call, and this only usage throws away the `err` value.

Let's just change that method to return a `bool` value `ok` (`true` is good) (and remove that error at all, as it only lives between these two calls).

Same applies to the other call commented below.

Reply:
@colega how does `makeLimitError` make an assumption? All it does is return `validationError`:

```go
return &validationError{
	err:  err,
	code: http.StatusBadRequest,
}
```

It could sound as if you're referring to `Limiter.FormatError`; it does handle four known error kinds, but it also has a `default` clause so it handles other kinds of errors also.

That said, I think your idea of returning a bool instead of an error is good.

Reply:
I will modify all 4 `AssertMaxXxx` methods (there are additional 2 that were initially not affected by this PR)

Reply:
Sorry @aknuds1 I was thinking about a different implementation I saw in a related PR.